### PR TITLE
feat: Add connection error handle

### PR DESCRIPTION
**Pull request by jiayisheji**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: [#1247](https://github.com/nestjs/mongoose/issues/1247)

## What is the new behavior?

Described here: [#1247](https://github.com/nestjs/mongoose/issues/1247)

## Does this PR introduce a breaking change?
- [x] Yes
- [ ] No

## Other information

### CODE REVIEW

File: lib/mongoose-core.module.ts [hunk 1]
```diff
@@ -39,11 +40,16 @@ export class MongooseCoreModule implements OnApplicationShutdown {
...
       connectionFactory,
+      connectionError,
       ...mongooseOptions
     } = options;
 
     const mongooseConnectionFactory =
       connectionFactory || ((connection) => connection);
+
+    const mongooseConnectionError =
+      connectionError || ((error) => {});
```
Comment (kamilmysliwiec) [inline lib/mongoose-core.module.ts:51]:
```suggestion
      connectionError || ((error) => error);   
```

File: lib/mongoose-core.module.ts [hunk 2]
```diff
@@ -59,7 +65,13 @@ export class MongooseCoreModule implements OnApplicationShutdown {
               await mongoose.createConnection(uri, mongooseOptions).asPromise(),
               mongooseConnectionName,
             ),
-          ).pipe(handleRetry(retryAttempts, retryDelay)),
+          ).pipe(
+            handleRetry(retryAttempts, retryDelay),
+            catchError((error) => {
+              mongooseConnectionError(error);
+              throw error;
```
Comment (kamilmysliwiec) [inline lib/mongoose-core.module.ts:72]:
```suggestion
              throw mongooseConnectionError(error);
```
what about this approach instead (this in combination with https://github.com/nestjs/mongoose/pull/1550/files#r1003070783)?

Comment (jiayisheji) [reply lib/mongoose-core.module.ts]: That's a very nice suggestion

File: lib/interfaces/mongoose-options.interface.ts [hunk 3]
```diff
@@ -9,6 +9,7 @@ export interface MongooseModuleOptions
   retryDelay?: number;
   connectionName?: string;
   connectionFactory?: (connection: any, name: string) => any;
+  connectionError?: (error: MongooseError) => MongooseError;
```
Comment (kamilmysliwiec) [inline lib/interfaces/mongoose-options.interface.ts:12]:
```suggestion
  connectionErrorFactory?: (error: MongooseError) => MongooseError;
```

File: lib/mongoose-core.module.ts [hunk 4]
```diff
@@ -39,11 +40,16 @@ export class MongooseCoreModule implements OnApplicationShutdown {
...
       connectionFactory,
+      connectionErrorFactory,
       ...mongooseOptions
     } = options;
 
     const mongooseConnectionFactory =
       connectionFactory || ((connection) => connection);
+
+    const mongooseConnectionError =
+    connectionErrorFactory || ((error) => error);
```
Comment (kamilmysliwiec) [inline lib/mongoose-core.module.ts:51]:
```suggestion
    const mongooseConnectionError =
      connectionErrorFactory || ((error) => error);   
```

**Comment by kamilmysliwiec** [comment]
lgtm

**Pull request merged**